README.md

# sidekiq-worker-killer
[![Gem Version](https://badge.fury.io/rb/sidekiq-worker-killer.svg)](https://badge.fury.io/rb/sidekiq-worker-killer)
[![CircleCI](https://circleci.com/gh/klaxit/sidekiq-worker-killer.svg?style=shield&circle-token=:circle-token)](https://circleci.com/gh/klaxit/sidekiq-worker-killer)

[Sidekiq](https://github.com/mperham/sidekiq) is probably the best background processing framework today. At the same time, memory leaks are very hard to tackle in Ruby and we often find ourselves with growing memory consumption. Instead of spending herculean effort fixing leaks, why not kill your processes when they got to be too large?

Highly inspired by [Gitlab Sidekiq MemoryKiller](https://gitlab.com/gitlab-org/gitlab-foss/-/blob/39c1731a53d1014eab7c876d70632b1abf738712/lib/gitlab/sidekiq_middleware/shutdown.rb) and [Noxa Sidekiq killer](https://github.com/Noxa/sidekiq-killer).

---
**NOTE**

This gem needs to get the used memory of the Sidekiq process. For
this we use [GetProcessGem](https://github.com/schneems/get_process_mem),
but be aware that if you are running Sidekiq in Heroku(or any container) the
memory usage will
[not be accurate](https://github.com/schneems/get_process_mem/issues/7).

---

quick-refs: [install](#install) | [usage](#usage) | [available options](#available-options) | [development](#development)

## Install
Use [Bundler](http://bundler.io/)
```ruby
gem "sidekiq-worker-killer"
```

## Usage

Add this to your Sidekiq configuration.

```ruby
require 'sidekiq/worker_killer'

Sidekiq.configure_server do |config|
  config.server_middleware do |chain|
    chain.add Sidekiq::WorkerKiller, max_rss: 480
  end
end
```

## Available options

The following options can be overridden.

| Option | Defaults | Description |
| ------- | ------- | ----------- |
| max_rss | 0 MB (disabled) | Max RSS in megabytes used by the Sidekiq process. Above this, shutdown will be triggered. |
| grace_time | 900 seconds | When shutdown is triggered, the Sidekiq process will not accept new job and wait at most 15 minutes for running jobs to finish. If Float::INFINITY specified, will wait forever. |
| shutdown_wait | 30 seconds | When the grace time expires, still running jobs get 30 seconds to stop. After that, kill signal is triggered. |
| kill_signal | SIGKILL | Signal to use to kill Sidekiq process if it doesn't stop. |
| gc | true | Try to run garbage collection before Sidekiq process stops in case of exceeded max_rss. |
| skip_shutdown_if | proc {false} | Executes a block of code after max_rss exceeds but before requesting shutdown. |
| on_shutdown | nil | Executes a block of code before just before requesting shutdown. This can be used to send custom logs or metrics to external services. |

*skip_shutdown_if* is expected to return anything other than `false` or `nil` to skip shutdown.

```ruby
require 'sidekiq/worker_killer'

Sidekiq.configure_server do |config|
  config.server_middleware do |chain|
    chain.add Sidekiq::WorkerKiller, max_rss: 480, skip_shutdown_if: ->(worker, job, queue) do
      worker.to_s == 'LongWorker'
    end
  end
end
```

## Development

Pull Requests are very welcome!

There are tasks that may help you along the way in a makefile:

```bash
make console # Loads the whole stack in an IRB session.
make test # Run tests.
make lint # Run rubocop linter.
```
Please make sure that you have tested your code carefully before opening a PR, and make sure as well that you have no style issues.

## Authors

See the list of [contributors](https://github.com/klaxit/sidekiq-worker-killer/contributors) who participated in this project.

## License

Please see [LICENSE](https://github.com/klaxit/sidekiq-worker-killer/blob/master/LICENSE)


CHANGELOG.md
**version 1.1.0**
- Add on_shutdown hook [#23] by [@Nirei]

**version 1.0.1**
- Support for Sidekiq >= 6.5.0. [#21] by [@bf4]

**version 1.0.0**
- Bump Sidekiq version to use the `Sidekiq::Process` API. [#12] by [@iGEL] and [#14] by [@pyrsmk]

**version 0.5.0**
- Option to skip shutdown on specific conditions (for example, for a specific job type). [#9] by [@msxavi]
- Ensure ProcessSet is up-to-date before checking for process that will need grace time. [#10] by [@msxavi]

**version 0.4.0**
- Fix grace time not getting honored. [#8] by [@VindictivePotato]
- Add gc option and support Float::INFINITY for grace_time option. [#4] by [@aishek]

**version 0.3.0**
- Run a full GC and then check memory again before initiating a restart. [#2] by [@BillFront]
- Don't wait for grace time if there's no work. [#1] by [@BillFront]

<!-- REFERENCES -->

[#1]: https://github.com/klaxit/sidekiq-worker-killer/pull/1
[#2]: https://github.com/klaxit/sidekiq-worker-killer/pull/2
[#4]: https://github.com/klaxit/sidekiq-worker-killer/pull/4
[#8]: https://github.com/klaxit/sidekiq-worker-killer/pull/8
[#9]: https://github.com/klaxit/sidekiq-worker-killer/pull/9
[#10]: https://github.com/klaxit/sidekiq-worker-killer/pull/10
[#12]: https://github.com/klaxit/sidekiq-worker-killer/pull/12
[#14]: https://github.com/klaxit/sidekiq-worker-killer/pull/14
[#21]: https://github.com/klaxit/sidekiq-worker-killer/pull/21
[#23]: https://github.com/klaxit/sidekiq-worker-killer/pull/23

[@aishek]: https://github.com/aishek
[@BillFront]: https://github.com/BillFront
[@iGEL]: https://github.com/iGEL
[@msxavi]: https://github.com/msxavi
[@pyrsmk]: https://github.com/pyrsmk
[@VindictivePotato]: https://github.com/VindictivePotato
[@bf4]: https://github.com/bf4
[@Nirei]: https://github.com/Nirei

spec/spec_helper.rb
$LOAD_PATH.unshift(File.dirname(__FILE__))
$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), "..", "lib"))

require "rspec"
require "sidekiq/worker_killer"

RSpec.configure do |config|
  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = "random"
end


spec/sidekiq/worker_killer_spec.rb
require "spec_helper"

describe Sidekiq::WorkerKiller do
  let(:sidekiq_process_set) { instance_double(Sidekiq::ProcessSet) }
  let(:sidekiq_process) { instance_double(Sidekiq::Process) }

  before do
    allow(subject).to receive(:warn) # silence "warn" logs
    allow(subject).to receive(:sleep) # reduces tests running time
    allow(sidekiq_process).to receive(:quiet!)
    allow(sidekiq_process).to receive(:stop!)
    allow(sidekiq_process).to receive(:stopping?)
  end

  describe "#sidekiq_process" do
    let(:identity) do
      "foobar"
    end
    let(:mock_processes) do
      [{
        "identity" => identity,
        "tag" => "baz",
        "started_at" => Time.now,
        "concurrency" => 5,
        "busy" => 2,
        "queues" => %w[low medium high]
      }]
    end
    it "finds the process by identity" do
      if subject.respond_to?(:config=)
        rspec_info "variant 6.5+: Sidekiq::ServerMiddleware"
        subject.config = mock_processes.first
      else
        rspec_info "variant pre 6.5: Sidekiq::Util"
        allow(subject).to receive(:identity).and_return(identity)
      end
      allow(Sidekiq).to receive(:redis)
      sidekiq_process_set = Sidekiq::ProcessSet.new
      sidekiq_mock_processes = mock_processes.map {|mock_process|
        Sidekiq::Process.new(mock_process)
      }
      allow(sidekiq_process_set).to receive(:each) {|&block|
        sidekiq_mock_processes.each(&block)
      }
      expect(sidekiq_process_set.to_a.map(&:identity)).to eq([identity])
      allow(Sidekiq::ProcessSet).to receive(:new).and_return(sidekiq_process_set)
      expect(subject.send(:identity)).to eq(identity)
      expect(subject.send(:sidekiq_process)).to be_a(Sidekiq::Process)
      expect(subject.send(:sidekiq_process).identity).to eq(identity)
    end
  end

  context "with stubbed sidekiq_process" do
    before do
      allow(Sidekiq::ProcessSet).to receive(:new) { sidekiq_process_set }
      allow(sidekiq_process_set).to receive(:find) { sidekiq_process }
      identity = "some-identity"
      if subject.respond_to?(:config=)
        rspec_info "variant 6.5+: Sidekiq::ServerMiddleware"
        subject.config = { "identity" => identity }
      else
        rspec_info "variant pre 6.5: Sidekiq::Util"
        allow(subject).to receive(:identity).and_return(identity)
      end
    end

    describe "#call" do
      let(:worker){ double("worker") }
      let(:job){ double("job") }
      let(:queue){ double("queue") }

      it "should yield" do
        expect { |b|
          subject.call(worker, job, queue, &b)
        }.to yield_with_no_args
      end

      context "when current rss is over max rss" do
        subject{ described_class.new(max_rss: 2) }

        before do
          allow(subject).to receive(:current_rss).and_return(3)
        end

        it "should request shutdown" do
          expect(subject).to receive(:request_shutdown)
          subject.call(worker, job, queue){}
        end

        it "should call garbage collect" do
          allow(subject).to receive(:request_shutdown)
          expect(GC).to receive(:start).with(full_mark: true, immediate_sweep: true)
          subject.call(worker, job, queue){}
        end

        context "and skip_shutdown_if is given" do
          subject{ described_class.new(max_rss: 2, skip_shutdown_if: skip_shutdown_proc) }

          context "and skip_shutdown_if is a proc" do
            let(:skip_shutdown_proc) { proc { |worker| true } }
            it "should NOT request shutdown" do
              expect(subject).not_to receive(:request_shutdown)
              subject.call(worker, job, queue){}
            end
          end

          context "and skip_shutdown_if is a lambda" do
            let(:skip_shutdown_proc) { ->(worker, job, queue) { true } }
            it "should NOT request shutdown" do
              expect(subject).not_to receive(:request_shutdown)
              subject.call(worker, job, queue){}
            end
          end

          context "and skip_shutdown_if returns false" do
            let(:skip_shutdown_proc) { proc { |worker, job, queue| false } }
            it "should still request shutdown" do
              expect(subject).to receive(:request_shutdown)
              subject.call(worker, job, queue){}
            end
          end

          context "and skip_shutdown_if returns nil" do
            let(:skip_shutdown_proc) { proc { |worker, job, queue| nil } }
            it "should still request shutdown" do
              expect(subject).to receive(:request_shutdown)
              subject.call(worker, job, queue){}
            end
          end
        end

        context "and on_shutdown is given" do
          subject{ described_class.new(max_rss: 2, on_shutdown: on_shutdown_proc) }

          context "and on_shutdown is a proc" do
            let(:on_shutdown_proc) { proc { |worker, job, queue| nil } }
            it "should execute on_shutdown hook" do
              expect(subject).to receive(:request_shutdown).once
              expect(on_shutdown_proc).to receive(:call).once
              subject.call(worker, job, queue){}
            end
          end

          context "and on_shutdown is a lambda" do
            let(:on_shutdown_proc) { ->(worker, job, queue) { nil } }
            it "should execute on_shutdown hook" do
              expect(subject).to receive(:request_shutdown).once
              expect(on_shutdown_proc).to receive(:call).once
              subject.call(worker, job, queue){}
            end
          end
        end

        context "when gc is false" do
          subject{ described_class.new(max_rss: 2, gc: false) }
          it "should not call garbage collect" do
            allow(subject).to receive(:request_shutdown)
            expect(GC).not_to receive(:start)
            subject.call(worker, job, queue){}
          end
        end

        context "but max rss is 0" do
          subject{ described_class.new(max_rss: 0) }
          it "should not request shutdown" do
            expect(subject).to_not receive(:request_shutdown)
            subject.call(worker, job, queue){}
          end
        end
      end
    end

    describe "#request_shutdown" do
      context "grace time is default" do
        before { allow(subject).to receive(:shutdown){ sleep 0.01 } }
        it "should call shutdown" do
          expect(subject).to receive(:shutdown)
          subject.send(:request_shutdown).join
        end
        it "should not call shutdown twice when called concurrently" do
          expect(subject).to receive(:shutdown).once
          2.times.map{ subject.send(:request_shutdown) }.each(&:join)
        end
      end

      context "grace time is 5 seconds" do
        subject{ described_class.new(max_rss: 2, grace_time: 5.0, shutdown_wait: 0) }
        it "should wait the specified grace time before calling shutdown" do
          # there are busy jobs that will not terminate within the grace time
          allow(subject).to receive(:no_jobs_on_quiet_processes?).and_return(false)

          shutdown_request_time     = nil
          shutdown_time             = nil

          # replace the original #request_shutdown to track
          # when the shutdown is requested
          original_request_shutdown = subject.method(:request_shutdown)
          allow(subject).to receive(:request_shutdown) do
            shutdown_request_time= Time.now
            original_request_shutdown.call
          end

          # track when the process has been required to stop
          expect(sidekiq_process).to receive(:stop!) do |*args|
            shutdown_time = Time.now
          end

          allow(Process).to receive(:kill)
          allow(Process).to receive(:pid).and_return(99)

          subject.send(:request_shutdown).join

          elapsed_time = shutdown_time - shutdown_request_time

          # the elapsed time between shutdown request and the actual
          # shutdown signal should be greater than the specified grace_time
          expect(elapsed_time).to be >= 5.0
        end
      end

      context "grace time is Float::INFINITY" do
        subject{ described_class.new(max_rss: 2, grace_time: Float::INFINITY, shutdown_wait: 0) }
        it "call signal only on jobs" do
          allow(subject).to receive(:jobs_finished?).and_return(true)
          allow(Process).to receive(:pid).and_return(99)
          expect(sidekiq_process).to receive(:quiet!)
          expect(sidekiq_process).to receive(:stop!)
          expect(Process).to receive(:kill).with('SIGKILL', 99)

          subject.send(:request_shutdown).join
        end
      end
    end
  end

  def rspec_info(msg)
    is_printing_progress = RSpec.configuration.formatters.any? { |f| /ProgressFormatter/.match?(f.class.name) }
    return if is_printing_progress
    RSpec.configuration.reporter.message msg
  end
end


lib/sidekiq/worker_killer.rb
require "get_process_mem"
require "sidekiq"
begin
  require "sidekiq/util"
  SidekiqComponent = Sidekiq::Util
rescue LoadError
  require "sidekiq/middleware/modules"
  SidekiqComponent = Sidekiq::ServerMiddleware
end
require "sidekiq/api"

# Sidekiq server middleware. Kill worker when the RSS memory exceeds limit
# after a given grace time.
class Sidekiq::WorkerKiller
  include SidekiqComponent

  MUTEX = Mutex.new

  # @param [Hash] options
  # @option options [Integer] max_rss
  #   Max RSS in MB. Above this, shutdown will be triggered.
  #   (default: `0` (disabled))
  # @option options [Integer] grace_time
  #   When shutdown is triggered, the Sidekiq process will not accept new job
  #   and wait at most 15 minutes for running jobs to finish.
  #   If Float::INFINITY is specified, will wait forever. (default: `900`)
  # @option options [Integer] shutdown_wait
  #   when the grace time expires, still running jobs get 30 seconds to
  #   stop. After that, kill signal is triggered. (default: `30`)
  # @option options [String] kill_signal
  #   Signal to use to kill Sidekiq process if it doesn't stop.
  #   (default: `"SIGKILL"`)
  # @option options [Boolean] gc
  #   Try to run garbage collection before Sidekiq process stops in case
  #   of exceeded max_rss. (default: `true`)
  # @option options [Proc] skip_shutdown_if
  #   Executes a block of code after max_rss exceeds but before requesting
  #   shutdown. (default: `proc {false}`)
  # @option options [Proc] on_shutdown
  #   Executes a block of code right before a shutdown happens.
  #   (default: `nil`)
  def initialize(options = {})
    @max_rss         = options.fetch(:max_rss, 0)
    @grace_time      = options.fetch(:grace_time, 15 * 60)
    @shutdown_wait   = options.fetch(:shutdown_wait, 30)
    @kill_signal     = options.fetch(:kill_signal, "SIGKILL")
    @gc              = options.fetch(:gc, true)
    @skip_shutdown   = options.fetch(:skip_shutdown_if, proc { false })
    @on_shutdown     = options.fetch(:on_shutdown, nil)
  end

  # @param [String, Class] worker_class
  #   the string or class of the worker class being enqueued
  # @param [Hash] job
  #   the full job payload
  #   @see https://github.com/mperham/sidekiq/wiki/Job-Format
  # @param [String] queue
  #   the name of the queue the job was pulled from
  # @yield the next middleware in the chain or the enqueuing of the job
  def call(worker, job, queue)
    yield
    # Skip if the max RSS is not exceeded
    return unless @max_rss > 0
    return unless current_rss > @max_rss
    GC.start(full_mark: true, immediate_sweep: true) if @gc
    return unless current_rss > @max_rss
    if skip_shutdown?(worker, job, queue)
      warn "current RSS #{current_rss} exceeds maximum RSS #{@max_rss}, " \
           "however shutdown will be ignored"
      return
    end

    warn "current RSS #{current_rss} of #{identity} exceeds " \
         "maximum RSS #{@max_rss}"

    run_shutdown_hook(worker, job, queue)
    request_shutdown
  end

  private

  def run_shutdown_hook(worker, job, queue)
    @on_shutdown.respond_to?(:call) && @on_shutdown.call(worker, job, queue)
  end

  def skip_shutdown?(worker, job, queue)
    @skip_shutdown.respond_to?(:call) && @skip_shutdown.call(worker, job, queue)
  end

  def request_shutdown
    # In another thread to allow underlying job to finish
    Thread.new do
      # Only if another thread is not already
      # shutting down the Sidekiq process
      shutdown if MUTEX.try_lock
    end
  end

  def shutdown
    warn "sending quiet to #{identity}"
    sidekiq_process.quiet!

    sleep(5) # gives Sidekiq API 5 seconds to update ProcessSet

    warn "shutting down #{identity} in #{@grace_time} seconds"
    wait_job_finish_in_grace_time

    warn "stopping #{identity}"
    sidekiq_process.stop!

    warn "waiting #{@shutdown_wait} seconds before sending " \
          "#{@kill_signal} to #{identity}"
    sleep(@shutdown_wait)

    warn "sending #{@kill_signal} to #{identity}"
    ::Process.kill(@kill_signal, ::Process.pid)
  end

  def wait_job_finish_in_grace_time
    start = Time.now
    sleep(1) until grace_time_exceeded?(start) || jobs_finished?
  end

  def grace_time_exceeded?(start)
    return false if @grace_time == Float::INFINITY

    start + @grace_time < Time.now
  end

  def jobs_finished?
    sidekiq_process.stopping? && sidekiq_process["busy"] == 0
  end

  def current_rss
    ::GetProcessMem.new.mb
  end

  def sidekiq_process
    Sidekiq::ProcessSet.new.find do |process|
      process["identity"] == identity
    end || raise("No sidekiq worker with identity #{identity} found")
  end

  def identity
    config[:identity] || config["identity"]
  end unless method_defined?(:identity)

  def warn(msg)
    Sidekiq.logger.warn(msg)
  end
end
